### PR TITLE
Dynamic programming/matrix chain multiplication

### PR DESCRIPTION
**Pull request by Shailaputri**
### Describe your change:

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms include at least one URL that points to Wikipedia or another similar explanation.
* [x] If this pull request resolves one or more open issues then the description above includes the issue number(s) with a [closing keyword](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue): "Fixes #ISSUE-NUMBER".

P.s. I would like to add that I started by trying to add doctests to matrix_chain_order.py script under Python/dynamic_programming/matrix_chain_order in order to improve its test coverage. However, I found the functions poorly written. Instead of changing the code, I have written a separate script with bug free code, problem explanation and required doctests. Moderators may consider deleting matrix_chain_order.py. Creating new PR by deleting #10549 and #10561.

### CODE REVIEW

**Comment by cclauss** [comment]
When would the algorithm be used in the real-world?
---

**Comment by Shailaputri** [comment]
Do I need to add a comment about real world applications in the docstring of this algorithm?

File: dynamic_programming/matrix_chain_multiplication.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+matrix_chain_multiply() returns an integer denoting
+minimum number of multiplications to multiply the chain.
+
+We do not need to perform actual multiplication here.
+We only need to decide the order in which to perform the multiplication.
+
+Hints:
+1. Number of multiplications (ie cost) to multiply 2 matrices
+of size m*p and p*n is m*p*n.
+2. Cost of matrix multiplication is neither associative ie (M1*M2)*M3 != M1*(M2*M3)
```
Comment (tianyizheng02) [inline dynamic_programming/matrix_chain_multiplication.py:31]: This is just incorrect. If the individual multiplications are well-defined, then matrix multiplication _is_ associative.

**Review by tianyizheng02 — CHANGES_REQUESTED** [review]
Do you have an implementation of the algorithm for determining the actual order of multiplications? Simply knowing _how many_ operations are needed isn't very useful and doesn't help with any of the real-world applications.

**Review by cclauss — APPROVED** [review]
The algorithm is framed in the same way on Wikipedia.  There are two different algorithms with benchmarks.

**Pull request merged**